File: src/main/java/com/wb/quasar/ActorTest.java
package com.wb.quasar;


import co.paralleluniverse.actors.*;
import co.paralleluniverse.fibers.SuspendExecution;
import co.paralleluniverse.strands.Strand;

import java.util.Objects;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;

public class ActorTest {
	public static void main(String[] args) throws Exception {
		run();
	}

	public static void run() {
		new NaiveActor("naive").spawn();
		try {
			Strand.sleep(Long.MAX_VALUE);
		} catch (SuspendExecution | InterruptedException suspendExecution) {
			suspendExecution.printStackTrace();
		}
	}

	@Upgrade
	static class BadActor extends BasicActor<String, Void> {
		private int count;

		@Override
		protected Void doRun() throws InterruptedException, SuspendExecution {
			System.out.println("(re)starting actor");
			for (; ; ) {
				String m = receive(300, TimeUnit.MILLISECONDS);
				if (m != null)
					System.out.println("Got a message: " + m);
				System.out.println("I am but a lowly actor that sometimes fails: - " + (count++));

				if (ThreadLocalRandom.current().nextInt(30) == 0)
					throw new RuntimeException("darn");

				checkCodeSwap(); // this is a convenient time for a code swap
			}
		}
	}

	static class NaiveActor extends BasicActor<Void, Void> {
		private ActorRef<String> myBadActor;

		public NaiveActor(String name) {
			super(name);
		}

		@Override
		protected Void doRun() throws InterruptedException, SuspendExecution {
			spawnBadActor();

			int count = 0;
			for (; ; ) {
				receive(500, TimeUnit.MILLISECONDS);
				myBadActor.send("hi from " + self() + " number " + (count++));
			}
		}

		private void spawnBadActor() {
			myBadActor = new BadActor().spawn();
			watch(myBadActor);
		}

		@Override
		protected Void handleLifecycleMessage(LifecycleMessage m) {
			if (m instanceof ExitMessage && Objects.equals(((ExitMessage) m).getActor(), myBadActor)) {
				System.out.println("My bad actor has just died of '" + ((ExitMessage) m).getCause() + "'. Restarting.");
				spawnBadActor();
			}
			return super.handleLifecycleMessage(m);
		}
	}
}

File: src/main/kotlin/ActorTest.kt
package demo

import co.paralleluniverse.actors.ActorRef
import co.paralleluniverse.actors.BasicActor
import co.paralleluniverse.actors.ExitMessage
import co.paralleluniverse.actors.LifecycleMessage
import co.paralleluniverse.fibers.Suspendable
import co.paralleluniverse.strands.Strand
import java.util.concurrent.ThreadLocalRandom
import java.util.concurrent.TimeUnit


fun main(args: Array<String>) {
    NaiveActor("naive").spawn()
    Strand.sleep(Long.MAX_VALUE)
}

class BadActor : BasicActor<String, Void>() {

    var count: Int = 0

    @Suspendable
    override fun doRun(): Void? {
        println("(re)starting actor")
        while (true) {
            var m: String? = receive(300, TimeUnit.MILLISECONDS)
            m?.run {
                println(" got a message $m")
            }
            println("I am but a lowly actor that sometime fails : - ${count++}")

            if (ThreadLocalRandom.current().nextInt(30) == 0)
                return null
//                throw RuntimeException("darn")
            checkCodeSwap()
        }

    }

}

class NaiveActor(val n: String) : BasicActor<Void, Void>(n) {
    lateinit private var myBadActor: ActorRef<String>

    @Suspendable
    override fun doRun(): Void? {
        spawnBadActor()
        var count = 0
        while (true) {
            receive(500, TimeUnit.MILLISECONDS)
            myBadActor.send(" hi from ${self()} number ${count++}")
        }
    }

    private fun spawnBadActor() {
        myBadActor = BadActor().spawn()
        watch(myBadActor)
    }

    override fun handleLifecycleMessage(m: LifecycleMessage?): Void? {
        if (m is ExitMessage && m.getActor() == myBadActor) {
            println(" my bad actor has just died of ${m.getCause()}. Restarting")
            spawnBadActor()
        }
        return super.handleLifecycleMessage(m)

    }

}